src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -52,7 +52,58 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 .span_label(span, format!("use of possibly uninitialized {}", item_msg))
                 .emit();
         } else {
-            let msg = ""; //FIXME: add "partially " or "collaterally "
+            // FIXME Should these be moved to impl Place?
+            fn is_deep_projection<'tcx>(
+                projected_place: &Place<'tcx>,
+                place: &Place<'tcx>
+            ) -> bool {
+                let mut cursor = projected_place;
+                let mut is_proj = false;
+                loop {
+                    if cursor == place {
+                        return is_proj;
+                    }
+                    let proj = match *cursor {
+                        Place::Local(..) | Place::Static(..) => return false,
+                        Place::Projection(ref proj) => proj
+                    };
+                    is_proj = true;
+                    cursor = &proj.base;
```

Comment:
I found this a bit confusing. We could code it up like this as an alternative:

```
place != projected_place && self.prefixes(projected_place, PrefixSet::All).any(|p| p == place)
```

though then you need the `MirBorrowckContext` for `self` (even though I think it doesn't get used when you have `PrefixSet::All`).

You could also add this method with a name like `place_extends(a, b)` (true if `a` is an extension of `b`) to `MirBorrowckContext`, so that the code below is like

```
if self.place_extends(foo, bar)
```

instead of 

```
is_deep_projection(foo, bar)
```